### PR TITLE
TW-281: Enable global search for recent chat, contacts

### PR DESCRIPTION
**Pull request by nqhhdev**
- Issue: #281 

- Resolved: 

1. First time go to the search screen, displays a list of recent chats (3).
2. When user searching, displays a list that includes the list of the most recent chats and contacts
3. User can click on an element of list then go to the conversation.

https://github.com/linagora/twake-on-matrix/assets/99852347/53396364-2ea0-47ec-b273-c655eeb567d4

### CODE REVIEW

File: lib/di/search/search_di.dart [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+  @override
+  String get scopeName => 'Search';
+
+  @override
+  void setUp(GetIt get) {
+    Logs().d('SearchDI::setUp()');
+
+    get.registerFactory<TomContactAPI>(() => TomContactAPI());
+
+    get.registerFactory<TomContactsDatasource>(() => TomContactsDatasourceImpl());
```
Comment (sherlockvn) [inline lib/di/search/search_di.dart:25]: I think registerSingleton is oke, no need for Factory

File: lib/di/search/search_di.dart [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+
+    get.registerFactory<ContactRepository>(() => TomContactRepositoryImpl());
```
Comment (sherlockvn) [inline lib/di/search/search_di.dart:27]: same

File: lib/di/search/search_di.dart [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+
+    get.registerFactory<LookupContactsInteractor>(() => LookupContactsInteractor());
```
Comment (sherlockvn) [inline lib/di/search/search_di.dart:29]: sanme

File: lib/di/search/search_di.dart [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+
+    get.registerFactory<FetchContactsInteractor>(() => FetchContactsInteractor());
```
Comment (sherlockvn) [inline lib/di/search/search_di.dart:31]: same for all

File: lib/domain/model/extensions/contact/presentation_contact_list_extension.dart [hunk 5]
```diff
@@ -0,0 +1,9 @@
+import 'package:fluffychat/domain/model/extensions/contact/presentation_contact_extension.dart';
+import 'package:fluffychat/presentation/model/presentation_contact.dart';
+import 'package:fluffychat/presentation/model/presentation_search.dart';
+
+extension PresentaionContactListExtension on List<PresentationContact> {
```
Comment (sherlockvn) [inline lib/domain/model/extensions/contact/presentation_contact_list_extension.dart:5]: not necessary

File: lib/domain/usecase/search/search_interactor.dart [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+    bool enableSearch = false
+  }) async* {
+    try {
+      if (enableSearch) {
+        final recentChat = await _searchRecentChat(context, keyword);
+        final contacts = await contactRepository.searchContact(query: ContactQuery(keyword: keyword), limit: limitContacts);
+
+        final presentationSearches = _comparePresentationSearches(
+          recentChat.toPresentationSearchList(context),
+          contacts.expand((contact) => contact.toPresentationContacts()).toList().toPresentationSearchList()
```
Comment (sherlockvn) [inline lib/domain/usecase/search/search_interactor.dart:38]: should separate to multiple lines

File: lib/domain/usecase/search/search_interactor.dart [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+        final rooms = await _getRecentChat(context, limitRecentChats: limitRecentChats);
+        yield Right(GetContactAndRecentChatSuccess(presentationSearches: rooms.toPresentationSearchList(context)));
+      }
+    } catch (e) {
+      yield Left(GetContactAndRecentChatFailed(exception: e));
+    }
+  }
+
+  Future<List<Room>> _getRecentChat(BuildContext context, {int? limitRecentChats}) async {
+    return Matrix.of(context).client.rooms.where((room) => !room.isSpace && !room.isStoryRoom)
```
Comment (sherlockvn) [inline lib/domain/usecase/search/search_interactor.dart:51]: Put `context` Which is belongs to Framework is not good in interactor. So i think you should pass Rooms or client to this method

File: lib/pages/chat_list/chat_list.dart [hunk 8]
```diff
@@ -135,7 +136,7 @@ class ChatListController extends State<ChatList>
     .client
     .rooms
     .where(getRoomFilterByActiveFilter(_activeFilterAllChats))
-    .where(isShowInChatList)
+    .where((room) => room.isShowInChatList())
```
Comment (sherlockvn) [inline lib/pages/chat_list/chat_list.dart:139]: why you change it back :))

File: lib/domain/usecase/search/search_interactor.dart [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+      .where((room) => room.isShowInChatList())
+      .take(limitRecentChats ?? 0)
+      .toList();
+  }
+
+  Future<List<Room>> _searchRecentChat(BuildContext context, String keyword) async {
+    return Matrix.of(context).client.rooms.where((room) => !room.isSpace && !room.isStoryRoom)
+      .where((room) => room.isShowInChatList())
+      .where((room) =>
```
Comment (dab246) [inline lib/domain/usecase/search/search_interactor.dart:60]: Why not combine 2 `where` conditions into one

File: lib/domain/model/extensions/contact/presentation_contact_list_extension.dart (hunk 5 above)
Comment (nqhhdev) [reply lib/domain/model/extensions/contact/presentation_contact_list_extension.dart:5]: - Why not necessary? 
It's converted `List<PresentationContact>` to `List<PresentationSearch>`

File: lib/domain/usecase/search/search_interactor.dart [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+        room.getLocalizedDisplayname(MatrixLocals(L10n.of(context)!))
+        .toLowerCase()
+        .contains(keyword.toLowerCase())
+    ).toList();
+  }
+
+  List<PresentationSearch> _comparePresentationSearches(List<PresentationSearch> recentChat, List<PresentationSearch> contacts) {
+    final isDuplicateElement = contacts.where((contact) {
```
Comment (dab246) [inline lib/domain/usecase/search/search_interactor.dart:68]: WDYT if use it like this.
`var mergedList = [...listX, ...listY, ...listZ].toSet(); `

File: lib/mixin/comparable_presentation_search_mixin.dart [hunk 11]
```diff
@@ -0,0 +1,24 @@
+import 'package:fluffychat/presentation/model/presentation_search.dart';
+
+class ComparablePresentationSearchMixin {
+  int comparePresentationSearch(PresentationSearch searchResultOne, PresentationSearch searchResultTwo) {
```
Comment (dab246) [inline lib/mixin/comparable_presentation_search_mixin.dart:4]: Somewhat confusing. Can you explain about it?

File: lib/pages/search/search_controller.dart [hunk 12]
```diff
@@ -0,0 +1,89 @@
...
+  String searchKeyword = "";
+
+  final SearchContactsAndRecentChatInteractor _searchContactsAndRecentChatInteractor = getIt.get<SearchContactsAndRecentChatInteractor>();
+  Debouncer<String>? _debouncer;
+  final TextEditingController textEditingController = TextEditingController();
+  StreamController<Either<Failure, GetContactAndRecentChatSuccess>> getContactAndRecentChatStream = StreamController();
+  void Function(String)? onSearchKeywordChanged;
+  late final isSearchModeNotifier = ValueNotifier(false);
+
+
```
Comment (dab246) [inline lib/pages/search/search_controller.dart:26]: remove line

File: lib/pages/search/search_controller.dart [hunk 13]
```diff
@@ -0,0 +1,89 @@
...
+  void init() {
+    _initializeDebouncer();
+    textEditingController.addListener(() {
+      if (textEditingController.text.isNotEmpty) {
+        isSearchModeNotifier.value = true;
+      } else {
+        isSearchModeNotifier.value = false;
+      }
```
Comment (dab246) [inline lib/pages/search/search_controller.dart:35]:
```suggestion
isSearchModeNotifier.value  = textEditingController.text.isNotEmpty;
```

File: lib/domain/usecase/search/search_interactor.dart (hunk 7 above)
Comment (nqhhdev) [reply lib/domain/usecase/search/search_interactor.dart:51]: Gooddddd

File: lib/domain/usecase/search/search_interactor.dart (hunk 10 above)
Comment (nqhhdev) [reply lib/domain/usecase/search/search_interactor.dart:68]: in your code I tried it and then result:

```
ListX = [e1, e2, e3]
ListY = [e4, e5, e6]

Noted: `e1.id == e4.id`

var mergedList = [...listX, ...listY].toSet().toList(); 

=> mergedList =  [e1, e2, e3, e4, e5, e6]
```

My code 

```
ListX = [e1, e2, e3]
ListY = [e4, e5, e6]

Noted: `e1.id == e4.id`

varr isDuplicateElement = ....(my code)

=> mergedList =  [e1, e2, e3, e5, e6]
```
IMO: I just want to create a list of duplicated elements and remove it from `mergedList`

File: lib/config/routes.dart [hunk 14]
```diff
@@ -172,6 +173,12 @@ class AppRoutes {
                     ),
                   ]
                 ),
+                VWidgetWithDependency(
```
Comment (hoangdat) [inline lib/config/routes.dart:177]: Why not a `VWidgetWithDependencies`?

File: lib/config/routes.dart [hunk 15]
```diff
@@ -172,6 +173,12 @@ class AppRoutes {
...
+                    di: ContactDI(),
+                    path: '/chatSearch',
+                    widget: const ChatSearch(),
```
Comment (hoangdat) [inline lib/config/routes.dart]: `Search`

File: lib/config/routes.dart [hunk 16]
```diff
@@ -172,6 +173,12 @@ class AppRoutes {
                     ),
                   ]
                 ),
+                VWidgetWithDependency(
+                    di: ContactDI(),
+                    path: '/chatSearch',
```
Comment (hoangdat) [inline lib/config/routes.dart]: `/search`

File: lib/config/routes.dart [hunk 17]
```diff
@@ -257,6 +264,12 @@ class AppRoutes {
                     ),
                   ],
                 ),
+                VWidgetWithDependency(
+                    di: ContactDI(),
+                    path: '/chatSearch',
+                    widget: const ChatSearch(),
+                    buildTransition: rightToLeftTransition
+                ),
```
Comment (hoangdat) [inline lib/config/routes.dart]: idem

File: lib/pages/chat_search/chat_recent_item_widget.dart [hunk 18]
```diff
@@ -0,0 +1,145 @@
...
+    this.room,
+    {
+      this.onTap,
+      Key? key,
+    }
+  ) : super(key: key);
+
+  @override
+  Widget build(BuildContext context) {
+    log(jsonEncode(room));
```
Comment (hoangdat) [inline lib/pages/chat_search/chat_recent_item_widget.dart]: remove it

File: lib/pages/chat_search/chat_recent_item_widget.dart [hunk 19]
```diff
@@ -0,0 +1,145 @@
...
+    if (directChatMatrixID == null) {
+      return _groupChatInformation(context);
+    } else {
+      return _directChatInformation(context);
+    }
+  }
+
+  Widget _groupChatInformation(BuildContext context) {
+    final actualMembersCount = (room.summary.mInvitedMemberCount ?? 0) +
+      (room.summary.mJoinedMemberCount ?? 0);
```
Comment (hoangdat) [inline lib/pages/chat_search/chat_recent_item_widget.dart]: deserve an extension for room

File: lib/pages/search/search.dart [hunk 20]
```diff
@@ -0,0 +1,82 @@
...
+  const Search({super.key});
+
+  @override
+  State<Search> createState() => SearchController();
+}
+
+class SearchController extends State<Search>  with ComparablePresentationContactMixin {
+
+  static const int limitRecentChats = 3;
+  static const int limitRecentContacts = 7;
```
Comment (hoangdat) [inline lib/pages/search/search.dart]:
```suggestion
  static const int limitPrefetchedRecentChats = 3;
  static const int limitPrefetchedRecentContacts = 7;
```

File: lib/presentation/model/presentation_search.dart [hunk 21]
```diff
@@ -0,0 +1,43 @@
+import 'package:equatable/equatable.dart';
+import 'package:matrix/matrix.dart';
+
+enum SearchTypeEnum {
+  contact,
+  recentChat,
+}
+
+class PresentationSearch extends Equatable {
```
Comment (hoangdat) [inline lib/presentation/model/presentation_search.dart:9]: `presentation contact` belong to `domain`, why this `presentation search` belong to `presentation`

File: lib/domain/usecase/search/search_interactor.dart (hunk 9 above)
Comment (hoangdat) [reply lib/domain/usecase/search/search_interactor.dart:63]: IMO, different should be better to read

File: lib/di/search/search_di.dart [hunk 22]
```diff
@@ -0,0 +1,39 @@
...
+class SearchDI extends BaseDI {
+
+  @override
+  String get scopeName => 'Search';
+
+  @override
+  void setUp(GetIt get) {
+    Logs().d('SearchDI::setUp()');
+
+    get.registerFactory<TomContactAPI>(() => TomContactAPI());
```
Comment (hoangdat) [inline lib/di/search/search_di.dart:23]: singleton

**Comment by hoangdat** [comment]
Change destination to `master`, keep the fixup commit. Thanks

**Comment by hoangdat** [comment]
@nqhhdev , I have new idea for the current contact
- only filter 7 chat in `chat list`
    - if every is direct chat, pick 7 latest chat
    - if also have group in chat list, get the member of group -> too find the 7 latest chat.

**Comment by nqhhdev** [comment]
> * if also have group in chat list, get the member of group -> too find the 7 latest chat.

I updated it, this is my demo

https://github.com/linagora/twake-on-matrix/assets/99852347/c39d7523-f512-40dc-bbe5-4c1456c38963

File: lib/domain/usecase/search/search_interactor.dart [hunk 23]
```diff
@@ -0,0 +1,79 @@
...
+
+  Stream<Either<Failure, GetContactAndRecentChatSuccess>> execute({
+    required BuildContext context,
+    required String keyword,
+    int? limitContacts,
+    int? limitRecentChats,
+    bool enableSearch = false
+  }) async* {
+    try {
+      if (enableSearch) {
```
Comment (hoangdat) [inline lib/domain/usecase/search/search_interactor.dart:29]: why not base on `keyword`?

File: lib/domain/usecase/search/search_interactor.dart [hunk 24]
```diff
@@ -0,0 +1,79 @@
...
+        final recentChat = await _searchRecentChat(context, keyword);
+        final contacts = await contactRepository.searchContact(query: ContactQuery(keyword: keyword), limit: limitContacts);
```
Comment (hoangdat) [inline lib/domain/usecase/search/search_interactor.dart:31]: did you support `load more`?

File: lib/domain/usecase/search/search_interactor.dart [hunk 25]
```diff
@@ -0,0 +1,79 @@
...
+
+  Future<List<Room>> _getRecentChat(BuildContext context, {int? limitRecentChats}) async {
+    return Matrix.of(context).client.rooms.where((room) => !room.isSpace && !room.isStoryRoom)
+      .where((room) => room.isShowInChatList())
+      .take(limitRecentChats ?? 0)
+      .toList();
+  }
+
+  Future<List<Room>> _searchRecentChat(BuildContext context, String keyword) async {
+    return Matrix.of(context).client.rooms.where((room) => !room.isSpace && !room.isStoryRoom)
```
Comment (hoangdat) [inline lib/domain/usecase/search/search_interactor.dart]: `(room) => !room.isSpace && !room.isStoryRoom` duplicated. Deserve a method

Comment (hoangdat) [reply lib/domain/usecase/search/search_interactor.dart]: What happens if we have a thousand of rooms?

File: lib/domain/usecase/search/search_interactor.dart (hunk 24 above)
Comment (nqhhdev) [reply lib/domain/usecase/search/search_interactor.dart:31]: We will update `load more` on this ticket #317

File: lib/domain/model/extensions/contact/presentation_contact_list_extension.dart [hunk 26]
```diff
@@ -0,0 +1,9 @@
+import 'package:fluffychat/domain/model/extensions/contact/presentation_contact_extension.dart';
+import 'package:fluffychat/presentation/model/presentation_contact.dart';
+import 'package:fluffychat/presentation/model/presentation_search.dart';
```
[the remark this answers is not in the document]
Comment (nqhhdev) [reply lib/domain/model/extensions/contact/presentation_contact_list_extension.dart:3]: I moved it to `presentation`

File: lib/domain/usecase/search/search_interactor.dart (hunk 25 above)
Comment (nqhhdev) [reply lib/domain/usecase/search/search_interactor.dart]: When we call anything from server or database there is always `limit`. I will update them by adding load more logic.

This ticket: #317

File: lib/domain/usecase/search/search_interactor.dart [hunk 27]
```diff
@@ -4,12 +4,12 @@ import 'package:fluffychat/di/global/get_it_initializer.dart';
 import 'package:fluffychat/domain/app_state/search/search_interactor_state.dart';
 import 'package:fluffychat/domain/model/contact/contact_query.dart';
 import 'package:fluffychat/domain/model/extensions/contact/contact_extension.dart';
-import 'package:fluffychat/domain/model/extensions/contact/presentation_contact_list_extension.dart';
+import 'package:fluffychat/presentation/extensions/contact/presentation_contact_list_extension.dart';
```
Comment (hoangdat) [inline lib/domain/usecase/search/search_interactor.dart:7]: I still see a lot of things in `presentation` here. It means that `domain` still know about `presentation`

File: lib/di/search/search_di.dart [hunk 28]
```diff
@@ -0,0 +1,39 @@
...
+
+    get.registerSingleton<TomContactsDatasource>(TomContactsDatasourceImpl());
+
+    get.registerSingleton<ContactRepository>(TomContactRepositoryImpl());
+
+    get.registerSingleton<LookupContactsInteractor>(LookupContactsInteractor());
+
+    get.registerSingleton<FetchContactsInteractor>(FetchContactsInteractor());
+
+    get.registerSingleton<LoadMoreInternalContacts>(LoadMoreInternalContacts());
```
Comment (hoangdat) [inline lib/di/search/search_di.dart:33]: why need to add it here? only add what you need?

File: lib/pages/search/search.dart [hunk 29]
```diff
@@ -61,19 +60,127 @@ class SearchController extends State<Search>  with ComparablePresentationContact
...
+
+      for (final user in users) {
+        final isDuplicateUser = recentChatListPresentationSearch
+          .any((existingUser) => existingUser.displayName == user.displayName);
+
+        if (!isDuplicateUser) {
+          recentChatListPresentationSearch.add(user);
+        }
+
+        if (recentChatListPresentationSearch.length == 6) {
```
Comment (sherlockvn) [inline lib/pages/search/search.dart:101]: should be a const for it

File: lib/pages/search/search.dart [hunk 30]
```diff
@@ -61,19 +60,127 @@ class SearchController extends State<Search>  with ComparablePresentationContact
...
   void initState() {
+    searchContactAndRecentChatController = SearchContactAndRecentChatController(context);
+    isSearchModeNotifier = ValueNotifier(false);
     listenContactsStartList();
+    listenSearchContactAndRecentChat();
     super.initState();
-    fetchContactsController.fetchCurrentTomContacts(limit: limitRecentContacts);
+    SchedulerBinding.instance.addPostFrameCallback((_) async {
+      if (mounted) {
+        searchContactAndRecentChatController?.init();
```
Comment (sherlockvn) [inline lib/pages/search/search.dart:171]: why put it in ScheduleBinding ?

File: lib/pages/search/search_controller.dart [hunk 31]
```diff
@@ -0,0 +1,91 @@
...
+    _debouncer?.values.listen((keyword) async {
+      Logs().d("SearchContactAndRecentChatController::_initializeDebouncer: searchKeyword: $searchKeyword");
+      searchKeyword = keyword;
+      if (onSearchKeywordChanged != null) {
+        onSearchKeywordChanged!(textEditingController.text);
+      }
+      final enableSearch = searchKeyword.isNotEmpty && searchKeyword != '';
+      fetchLookupContacts(
+        enableSearch: enableSearch,
+        limitRecentChats: !enableSearch ? 3 : null
```
Comment (sherlockvn) [inline lib/pages/search/search_controller.dart:52]: const it

File: lib/presentation/mixin/load_more_search_mixin.dart [hunk 32]
```diff
@@ -0,0 +1,30 @@
...
+
+  final haveMoreCountactsNotifier = ValueNotifier(true);
+
+  final lastContactIndexNotifier = ValueNotifier(0);
+
+  List<PresentationSearch> oldContactList = [];
+  List<PresentationSearch> oldContactAndRecentChatList = [];
+
+  void listenForScrollChanged({required FetchContactsController fetchContactsController}) {
+    scrollController.addListener(() {});
```
Comment (sherlockvn) [inline lib/presentation/mixin/load_more_search_mixin.dart:16]: what does it do ?

File: lib/domain/model/search/search_model.dart [hunk 33]
```diff
@@ -0,0 +1,41 @@
+import 'package:equatable/equatable.dart';
+import 'package:matrix/matrix.dart';
+
+enum SearchElementTypeEnum {
+  contact,
+  recentChat,
+}
+
+class Search extends Equatable {
```
Comment (hoangdat) [inline lib/domain/model/search/search_model.dart:9]: `SearchModel` or `SearchResult`

File: lib/presentation/mixin/load_more_search_mixin.dart (hunk 32 above)
Comment (nqhhdev) [reply lib/presentation/mixin/load_more_search_mixin.dart:16]: I will update them by adding load more logic.

This ticket: https://github.com/linagora/twake-on-matrix/issues/317

File: lib/pages/search/search.dart (hunk 30 above)
Comment (nqhhdev) [reply lib/pages/search/search.dart:171]:
```
  void _getContactAndRecentChat() {
    _searchContactsAndRecentChatInteractor.execute(
      rooms: Matrix.of(context).client.rooms,
      matrixLocalizations: MatrixLocals(L10n.of(context)!),
      keyword: '',
      limitRecentChats: limitPrefetchedRecentChats,
      limitContacts: limitSearchingPrefetchedRecentContacts,
    ).listen((event) {
      getContactAndRecentChatStream.add(event);
    });
  }
```

Currently I use `context` when initState so I need to put it in ScheduleBinding to make sure the context is created first.

**Review by sherlockvn — APPROVED** [review]

File: lib/domain/model/room/room_extension.dart [hunk 34]
```diff
@@ -0,0 +1,40 @@
+
+import 'package:fluffychat/domain/model/search/search_model.dart';
+import 'package:fluffychat/utils/matrix_sdk_extensions/client_stories_extension.dart';
+import 'package:matrix/matrix.dart';
+
+extension RoomExtension on Room {
```
Comment (sherlockvn) [inline lib/domain/model/room/room_extension.dart:6]: please use anotherName, more specific, because there are RoomExtension in utils folder

Comment (nqhhdev) [reply lib/domain/model/room/room_extension.dart:6]: Don't worry, it's only one `RoomExtension `

**Review by hoangdat — APPROVED** [review]

File: lib/mixin/comparable_presentation_search_mixin.dart (hunk 11 above)
Comment (nqhhdev) [reply lib/mixin/comparable_presentation_search_mixin.dart:5]: After concatenating the properties into strings, it compares the two strings using the compareTo method, which returns an integer representing the comparison result.

The method returns the result of the string comparison, which can be negative if bufferOne is lexicographically less than bufferTwo, zero if they are equal, or positive if bufferOne is lexicographically greater than bufferTwo.

**Pull request merged**